docs/design/partitioning/07-parallel-migrations.md

```diff
@@ -0,0 +1,536 @@
+|-------------|
+
+## Background
+
+The partition table is an eventually consistent and monotonic data structure. It’s only updated by the master member 
+and published to the cluster. Some members may lag behind the master’s version at any time but for those members to be 
+able to participate in a migration, they have to get the latest partition table from master. To provide the monotonic 
+order, the partition table has a global version. And because of this global order, we have to migrate partitions 
+one-by-one. Once the master member changes, it fetches the partition tables and the list of completed & active migrations 
+from the cluster and determines the most recent version of the partition table and then publishes it back. 
+
+Right now this is a very strict guarantee, because there were data loss issues during shutdown and migrations 
+a few years back and we wanted all partition updates and migrations to be more consistent.
+
+The good thing is, in practice, we don’t need this global partition table update order. Because, partitions are not 
+dependent on each other. Instead of a global partition table version, we can have a version per partition. And we can 
+update partitions independently. So, after each partition migration, we can just publish that partition info only, 
+not the whole partition table. Also we can migrate different partitions concurrently. For example, assume there are 10 
+members, we can execute 5 migrations between 5 member pairs in parallel. With these two optimizations 
+(removal of global update order and parallel migrations), repartitioning can become significantly faster than current.
+
+### Previous Work
+
+Previous works related to partitioning system and migration mechanism: 
+
+| Version | Design |
+| :-----: | :----- |
+| 3.7 | [Avoid Data Loss on Migration](01-avoid-data-loss-on-migration.md) |
+| 3.7 | [Graceful Shutdown Improvements](02-graceful-shutdown-improvements.md) |
+| 3.9 | [Fine-Grained Anti Entropy Mechanism](03-fine-grained-anti-entropy-mechanism.md) |
+| 3.9 | [Gigantic Cache Migration Enhancements](04-gigantic-cache-migration-enhancements.md) |
+| 3.12 | [Speed-Up Partition Migrations](05-speed-up-partition-migrations.md) |
+| 4.0 | [Refactor MigrationListener API](06-refactor-migrationlistener-api.md) |
+
+_Being familiar with these will help to understand the new changes._
+ 
+## Design
+
+### 1 - Global Version to Partition Level
+
+First of the major changes is the removal of the global partition table update ordering. Instead we will version each 
+partition individually. 
+
+#### 1.1 - Current Versioning Mechanism
+
+Currently, master member increments global version on each partition update. When the partition table update is received 
+from the master, members decide whether to apply or to ignore the update comparing their local version to received 
+version. If local version is less than received version then partition table is applied, otherwise it's ignored. 
```

Comment:
maybe we can reference all-checks-to-start-a-migration-operation documentation, if there is any.

Reply:
Afaik, we haven't documented all of these verifications before. Even some of them are introduced eventually with bug-fixing.